src/core/Fingerprint.cpp

```diff
@@ -0,0 +1,237 @@
+
+    releaseDevice();
+    return true;
+}
+
+void CFingerprint::registerSleepHandler() {
+    m_sDBUSState.login = sdbus::createProxy(*m_sDBUSState.connection.get(), "org.freedesktop.login1", "/org/freedesktop/login1");
+
+    m_sDBUSState.login->uponSignal("PrepareForSleep").onInterface(LOGIN_MANAGER).call([this](bool start) {
+        Debug::log(LOG, "prepare for sleep");
+        if (start) {
+            m_sDBUSState.sleeping = true;
+            stopVerify();
+            m_sDBUSState.inhibitLock.reset();
+        } else {
+            m_sDBUSState.sleeping = false;
+            inhibitSleep();
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+            startVerify();
+        }
+    });
+    m_sDBUSState.login->finishRegistration();
+    inhibitSleep();
+}
+
+void CFingerprint::inhibitSleep() {
+    m_sDBUSState.login->callMethod("Inhibit")
+        .onInterface(LOGIN_MANAGER)
+        .withArguments("sleep", "hyprlock", "Release fingerprint before sleep", "delay")
+        .storeResultsTo(m_sDBUSState.inhibitLock);
+}
+
+std::optional<sdbus::Error> CFingerprint::createDeviceProxy() {
+    auto              proxy = sdbus::createProxy(*m_sDBUSState.connection.get(), FPRINT, "/net/reactivated/Fprint/Manager");
+
+    sdbus::ObjectPath path;
+    try {
+        proxy->callMethod("GetDefaultDevice").onInterface(MANAGER).storeResultsTo(path);
+    } catch (sdbus::Error& e) {
+        Debug::log(WARN, "Couldn't connect to Fprint service ({})", e.what());
+        return e;
+    }
+    Debug::log(LOG, "Using dbus path {}", path.c_str());
+    m_sDBUSState.device = sdbus::createProxy(*m_sDBUSState.connection.get(), FPRINT, path);
+
+    m_sDBUSState.device->uponSignal("VerifyFingerSelected").onInterface(DEVICE).call([](const std::string& finger) { Debug::log(LOG, "Finger: {}", finger); });
+    m_sDBUSState.device->uponSignal("VerifyStatus").onInterface(DEVICE).call([this](const std::string& result, const bool done) { handleVerifyStatus(result, done); });
+
+    m_sDBUSState.device->finishRegistration();
+    return {};
+}
+
+void CFingerprint::handleVerifyStatus(const std::string& result, bool done) {
+    Debug::log(LOG, "Status: {} {}", done, result);
+    auto matchResult = s_mapStringToTestType[result];
+    if (m_sDBUSState.sleeping && matchResult != MatchResult::Disconnected) {
+        return;
+    }
+    std::unique_lock<std::mutex> lk(m_sDBUSState.failTextMutex);
+    switch (matchResult) {
+        case MatchResult::NoMatch: //The verification did not match, Device.VerifyStop should now be called.
+            stopVerify();
+            if (m_sDBUSState.retries < 3) {
+                startVerify();
+                done = false;
+                m_sDBUSState.retries++;
+                m_sDBUSState.failText = RETRY_MESSAGE;
+                if (clearRetryTextTimer) clearRetryTextTimer->cancel();
+                clearRetryTextTimer   = g_pHyprlock->addTimer(
+                    std::chrono::seconds(2),
```

Comment:
This could just be the `input-field:fail_timeout` option I think?

Reply:
It's not straightforward AFAICT since input-field is a special category according to the config so I'm not sure how to get it aside from looping through the widget configs and finding the input-field, if it's there, but let me know if I'm missing something.

Reply:
You are right.

Idea: Add `$FPRINTPROMPT` and `$FPRINTFAIL` like `$PROMPT` and `$FAIL` in `Auth.cpp` (`getLastPrompt` and `getLastFailText`), then we would not need the timer at all. Although you would not be able to have `$FPRINTFEEDBACK` like behavior. Currently you also don't have that for normal auth. Switching between fail and prompt is input-field exclusive.

Reply:
There's a signal for finger presence, so I switched to change the $FPRINTMESSAGE at that time, removing the need for the timer.